homeassistant/components/radarr/config_flow.py

```diff
@@ -0,0 +1,147 @@
+            except (ClientConnectorError, exceptions.ArrConnectionException):
+                errors = {"base": "cannot_connect"}
+            except exceptions.ArrException:
+                errors = {"base": "unknown"}
+            if not errors:
+                if self.entry:
+                    self.hass.config_entries.async_update_entry(
+                        self.entry, data=user_input
+                    )
+                    await self.hass.config_entries.async_reload(self.entry.entry_id)
+
+                    return self.async_abort(reason="reauth_successful")
+
+                return self.async_create_entry(
+                    title=DEFAULT_NAME,
+                    data=user_input,
+                )
+
+        user_input = user_input or {}
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(
+                        CONF_URL, default=user_input.get(CONF_URL, DEFAULT_URL)
+                    ): str,
+                    vol.Optional(CONF_API_KEY): str,
+                    vol.Optional(
+                        CONF_VERIFY_SSL,
+                        default=user_input.get(CONF_VERIFY_SSL, False),
+                    ): bool,
+                }
+            ),
+            errors=errors,
+        )
+
+    async def async_step_import(self, config: dict[str, Any]) -> FlowResult:
+        """Import a config entry from configuration.yaml."""
+        for entry in self._async_current_entries():
+            if entry.data[CONF_API_KEY] == config[CONF_API_KEY]:
+                _part = config[CONF_API_KEY][0:4]
+                _msg = f"Radarr yaml config with partial key {_part} has been imported. Please remove it"
+                LOGGER.warning(_msg)
+                return self.async_abort(reason="already_configured")
+        proto = "https" if config[CONF_SSL] else "http"
+        host_port = f"{config[CONF_HOST]}:{config[CONF_PORT]}"
+        path = ""
+        if config["urlbase"].rstrip("/") not in ("", "/", "/api"):
+            path = config["urlbase"].rstrip("/")
+        return self.async_create_entry(
+            title=DEFAULT_NAME,
+            data={
+                CONF_URL: f"{proto}://{host_port}{path}",
+                CONF_API_KEY: config[CONF_API_KEY],
+                CONF_VERIFY_SSL: False,
+            },
+        )
+
+
+async def validate_input(
+    hass: HomeAssistant, data: dict[str, Any]
+) -> tuple[str, str, str] | str | None:
+    """Validate the user input allows us to connect."""
+    host_configuration = PyArrHostConfiguration(
+        api_token=data.get(CONF_API_KEY, ""),
+        verify_ssl=data[CONF_VERIFY_SSL],
+        url=data[CONF_URL],
+    )
+    radarr = RadarrClient(
+        host_configuration=host_configuration,
+        session=async_get_clientsession(hass),
+    )
+    if CONF_API_KEY not in data:
+        return await radarr.async_try_zeroconf()
```

Comment:
We shouldn't return a string directly from the library to be used as an error string. The integration should control that.